### ISSUE TITLE
Production build stuck waiting for MySQL to initialize (GKE & kubernetes)

### ISSUE THREAD

**Issue by eurobob** [visiblevc/wordpress-starter#52]
## Overview

I'm trying to deploy this image with Google Container Engine (https://cloud.google.com/container-engine/docs/tutorials/hello-wordpress), but it seems to get stuck on initialising MySQL.

## `kubernetes pod log` output

```
wordpress_1  | ===============================================================================
wordpress_1  |                          Begin WordPress Installation
wordpress_1  | ===============================================================================
wordpress_1  | ==> Installing wordpress
wordpress_1  |   -> Downloading... ✓
wordpress_1  | ==> Waiting for MySQL to initialize...
```

**Comment by dsifford** [comment]
This image requires a MySQL or MariaDB container be running and available ('exposed') prior to starting up. I don't have much experience with kubernetes, but in the example that you link to, I assume the tutum/wordpress image does not depend on a running database prior to running and instead can be linked to one at some other point after starting up.

**Comment by eurobob** [comment]
Makes sense. To be honest I'm not planning on using MySQL on each machine. Since it will be an autoscaling group I will be using the Google Cloud SQL service. Do you have any suggestions on how I would bypass the MySQL setup? I assume I would just change the environment variables in the docker-compose.yml?

**Comment by eurobob** [comment]
I tried removing the section from run.sh and rebuilding my docker image, but it didn't seem to work :/

**Comment by dsifford** [comment]
@eurobob Not sure if I'm understanding what you're trying to do. Can you clarify?

**Option 1**: You're trying to cluster several sites around a single database container

**Option 2:** You're trying to cluster several singular WordPress instances around a single database container

## If option 1
This project is not currently set up to be able to do something like that. If you don't mind me asking, why is that something that you'd even want to do? You'd be creating a single point of failure across all the sites!

## If option 2
I'm not sure there's an easy way to go about doing that since wordpress stores runtime files in the local filesystem. For example, if you have 2 instances running and one of your admins gets routed to instance 2 and decides to update or install a plugin, instance 1 will have no idea that that happened (since plugins are installed directly to wp-content). The same result is true if an admin decides to upload any type of media. So you'll quickly go out of sync.

I'm sure it's possible with elbow grease, but it won't be easy. You'd have to offload all your media to some third-party service (like S3), and figure out a way to maintain synchronization between your instances for plugin/theme updates.

A quick google search turned this blog post up which might be helpful: http://blog.celingest.com/en/2014/10/16/new-tutorial-highly-available-wordpress-part-i/

That all said, if you are aware of an easier way of going about this, or if you decide you still want to figure it out and take the time to do it, I definitely want to hear about it. Keep me posted.

## TL;DR 

### Option 1
This configuration will not work without a moderate amount of modification. I wouldn't recommend doing that either because it would be at the cost of one of Docker's single-greatest selling points; separation of concerns.

### Option 2
This configuration will not work for something like that. If you take the time to put something like that together, I want to hear about it.

**Comment by eurobob** [comment]
Honestly I'm a little confused as to what i'm doing.

I just want to deploy a wordpress app on an autoscaling cloud infrastructure. Maybe there is a simple way that I just can't seem to find documented well. It felt like I had to go through all this docker stuff based on what everyone talks about these days. I tried AppEngine already but the performance just wasn't good enough for me.

I realise this is out of the scope of this "issue" now, so i'll close it. FML

**Comment by dsifford** [comment]
@eurobob Happy to help point you in the right direction if I can.

WordPress is just inherently non-scalable by design. 

If you simply want to be able to set up a few cloud WordPress sites, this project makes that a breeze. You just need to use docker compose rather than kubernetes or swarm (since we're not clustering).

**Comment by eurobob** [comment]
Thanks a lot, i've definitely been experiencing the non-scalable irks. I have joined a company and inherited the application. It's the kind of thing that definitely shouldn't have been built on top of wordpress so it's very fragile anyway, but it works and we're doing a huge marketing push at the start of 2017 so we need to make sure it handles any traffic surges. Then I can get started on a much better v2.

So I was hoping i'd be able to deploy the basic WP site (happy to version all plugins etc.), and then use Cloud storage & SQL to share uploads and database across all instances. Just looking for the absolute quickest and easiest way to do this so I can forget about it and tend to more important matters 😄

**Comment by dsifford** [comment]
@eurobob Look into [easyengine](https://github.com/EasyEngine/easyengine). It's not a docker project, but it's easily the most performant implementation of WordPress that I'm aware of. 

Installation is brainless and they're working on fully integrating with WP-CLI.

I made a docker implementation with easyengine (which is now currently broken due to an API change) and was able to get a 99% pagespeed. Using a standard apache wordpress installation with everything else the exact same, the pagespeed was around 70%.

For giggles, I also load tested easyengine on one of my high-traffic sites back in July and was able to get these results (don't have a current apache configuration for comparison, but there's no way it would come close to these):

**Configuration:**
- 10,000 clients/second
- Timeout: 10 seconds

![image](https://cloud.githubusercontent.com/assets/5240018/20417354/a931d20a-ad12-11e6-834f-d7938befdab7.png)

![image](https://cloud.githubusercontent.com/assets/5240018/20417362/b24b6e6e-ad12-11e6-9812-a795a4dfae09.png)

![image](https://cloud.githubusercontent.com/assets/5240018/20417367/bcee6006-ad12-11e6-8566-d682b7cb0ee4.png)

Quite a bit of timeouts, but no errors.

Hope that helps.

**Comment by eurobob** [comment]
Awesome! Thanks so much for this. I used it to set up an image template for Google Compute Engine and set up my load balancer along with Cloud SQL, Cloud Storage & Data store for sessions.

However, I'm seeing very poor performance. I'm using a single CPU, 3.75GB ram machine for my template. I did an Apache Bench test and got these results:

`$ ab -n 100 -c 10 https://XXX.XXX.XXX.XXX/

Server Software:        nginx
Server Hostname:        XXX.XXX.XXX.XXX
Server Port:            443
SSL/TLS Protocol:       TLSv1.2,ECDHE-RSA-CHACHA20-POLY1305,2048,256

Document Path:          /
Document Length:        104331 bytes

Concurrency Level:      10
Time taken for tests:   47.532 seconds
Complete requests:      100
Failed requests:        0
Total transferred:      10492796 bytes
HTML transferred:       10433100 bytes
Requests per second:    2.10 [#/sec] (mean)
Time per request:       4753.178 [ms] (mean)
Time per request:       475.318 [ms] (mean, across all concurrent requests)
Transfer rate:          215.58 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:      134  168  63.8    150     580
Processing:   992 4409 2156.7   5056    8352
Waiting:      666 2406 1034.3   2663    4537
Total:       1146 4577 2172.5   5209    8519

Percentage of the requests served within a certain time (ms)
  50%   5209
  66%   5924
  75%   6297
  80%   6642
  90%   7242
  95%   7943
  98%   8482
  99%   8519
 100%   8519 (longest request)`

So i checked out the running processes on one of the instances and found a lot of php-fpm: pool www processes that are consuming 30% CPU while it was idle. Here are the processes during the Apache Bench:

<img width="705" alt="screen shot 2016-11-21 at 15 54 24" src="https://cloud.githubusercontent.com/assets/4255350/20474924/fa0f6aae-b003-11e6-8300-f4d215887fe7.png">


84.5%!!! This can't be normal surely? It's a wordpress home page. It does use visual composer & advanced custom fields, and in general uses a LOT of plugins. Could that be why the instances are getting hammered?

I understand we've departed the original issue a lot but any insight would be very helpful, particularly in how I would manage to debug this mess. Thanks :)

**Comment by eurobob** [comment]
So I deactivated all the plugins and the php-fpm usage dropped to 3%. I guess it really is a piece of shit app

**Comment by dsifford** [comment]
Which plugin was causing the issue @eurobob

**Comment by eurobob** [comment]
@dsifford The culprit was Advanced Custom Fields, but once I checked the access logs I realised that it was the Google Health Check running every 5 seconds that was causing it to overload the server. Changed that to 60s and now my instance is idling at <3% CPU. 😄

Thanks for letting me know about easyengine! Hopefully this will help someone in the future stumbling upon this thread.

**Comment by dsifford** [comment]
@eurobob No problem.

Not sure if I'm understanding you correctly. Are you saying that it is, in fact, a codebase issue with ACF that causes the rapidfire health checks? Should @elliotcondon be aware of this?

No problem. Happy to help.

**Comment by eurobob** [comment]
I'm not sure if there is a codebase issue. I just think that my instance capacity was unable to adequately cope with running through ACF every 5 seconds.

However, when I profiled the installation with the P3 plugin I did notice that ACF does contribute to ~20% of my page load time. I assumed this was to be expected as part of the wordpress plugin bloat? Otherwise maybe there are some improvements that can be made in which case i'd be glad to help out with the assessment.
